// @rooks/use-boundingclientrect/lib/index.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('react')) :
  typeof define === 'function' && define.amd ? define(['react'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, global.useBoundingclientrect = factory(global.React));
}(this, (function (react) { 'use strict';

  var config = {
      attributes: true,
      characterData: true,
      subtree: true,
      childList: true
  };
  /**
   *
   * useMutationObserver hook
   *
   * Returns a mutation observer for a React Ref and fires a callback
   *
   * @param {MutableRefObject<HTMLElement | null>} ref React ref on which mutations are to be observed
   * @param {MutationCallback} callback Function that needs to be fired on mutation
   * @param {MutationObserverInit} options
   */
  function useMutationObserver(ref, callback, options = config) {
      react.useEffect(() => {
          // Create an observer instance linked to the callback function
          if (ref.current) {
              const observer = new MutationObserver(callback);
              // Start observing the target node for configured mutations
              observer.observe(ref.current, options);
              return () => {
                  observer.disconnect();
              };
          }
      }, [callback, options]);
  }

  /**
   * useDidMount hook
   * Calls a function on mount
   *
   * @param {function} callback Callback function to be called on mount
   */
  function useDidMount(callback) {
      react.useEffect(() => {
          if (typeof callback === "function") {
              callback();
          }
      }, []);
  }

  /**
   * @param element HTML element whose boundingclientrect is needed
   * @return ClientRect
   */
  function getBoundingClientRect(element) {
      return element.getBoundingClientRect();
  }
  /**
   * useBoundingclientRect hook
   * @param ref The React ref whose ClientRect is needed
   * @return ClientRect
   */
  function useBoundingclientrect(ref) {
      const [value, setValue] = react.useState(null);
      const update = react.useCallback(() => {
          setValue(ref.current ? getBoundingClientRect(ref.current) : null);
      }, []);
      useDidMount(() => {
          update();
      });
      useMutationObserver(ref, update);
      return value;
  }

  return useBoundingclientrect;

})));
